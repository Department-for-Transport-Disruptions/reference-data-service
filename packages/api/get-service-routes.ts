import { APIGatewayEvent, APIGatewayProxyResultV2 } from "aws-lambda";
import { getServiceStops, ServiceStop, ServiceStops, ServiceStopsQueryInput } from "./client";
import { ClientError } from "./error";
import { executeClient } from "./execute-client";
import { flattenStops } from "./get-service-stops";

export const main = async (event: APIGatewayEvent): Promise<APIGatewayProxyResultV2> =>
    executeClient(event, getQueryInput, getServiceStops, formatStopsRoutes);

export const getQueryInput = (event: APIGatewayEvent): ServiceStopsQueryInput => {
    const { pathParameters } = event;

    const serviceId = pathParameters?.serviceId;

    if (!serviceId) {
        throw new ClientError("Service ID must be provided");
    }

    if (isNaN(Number(serviceId))) {
        throw new ClientError("Provided service ID is not valid");
    }

    return {
        serviceId: Number(serviceId),
    };
};

const filterStops = (flattenedStops: ServiceStop[], direction: string) => {
    return (
        flattenedStops
            // filter stops by direction
            .filter((stop) => stop.direction === direction)
            // remove any duplicates on atcoCode and sequence number
            .filter(
                (stop, index, self) =>
                    self.findIndex(
                        (other) => stop.atcoCode === other.atcoCode && stop.sequenceNumber === other.sequenceNumber,
                    ) === index,
            )
            // sort stops by sequence number
            .sort((stop, other) => Number(stop.sequenceNumber) - Number(other.sequenceNumber))
            // remove duplicate adjacent stops
            .filter((stop, i) => (i > 0 ? stop.atcoCode !== flattenedStops[i - 1].atcoCode : true))
    );
};

export const formatStopsRoutes = async (
    stops: ServiceStops,
    // eslint-disable-next-line @typescript-eslint/require-await
): Promise<{ outbound: ServiceStop[]; inbound: ServiceStop[] }> => {
    const flattenedStops = flattenStops(stops);
    const outbound = filterStops(flattenedStops, "outbound");
    const inbound = filterStops(flattenedStops, "inbound");
    return { outbound, inbound };
};
