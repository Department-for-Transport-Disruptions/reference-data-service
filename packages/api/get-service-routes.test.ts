import { APIGatewayEvent } from "aws-lambda";
import { describe, expect, it } from "vitest";
import { formatStopsRoutes, getQueryInput } from "./get-service-routes";
import { stopsDbData } from "./test/testdata";
import { BusStopType, ServiceStop, VehicleMode } from "./client";

describe("get-service-routes", () => {
    describe("input generation", () => {
        it("handles serviceId", () => {
            const event = {
                pathParameters: {
                    serviceId: "234",
                },
            } as unknown as APIGatewayEvent;

            expect(getQueryInput(event)).toEqual({ serviceId: 234 });
        });

        it("throws a ClientError if no serviceId provided", () => {
            const event = {
                pathParameters: {},
            } as unknown as APIGatewayEvent;

            expect(() => getQueryInput(event)).toThrowError("Service ID must be provided");
        });

        it("handles serviceId, stopTypes, busStopType and modes", () => {
            const event = {
                pathParameters: {
                    serviceId: "234",
                    stopTypes: "BCT",
                    busStopType: BusStopType.mkd,
                    modes: VehicleMode.bus,
                },
            } as unknown as APIGatewayEvent;

            expect(getQueryInput(event)).toEqual({
                serviceId: 234,
                stopTypes: ["BCT"],
                busStopType: [BusStopType.mkd],
                modes: [VehicleMode.bus],
            });
        });

        it("throws a ClientError if invalid serviceId provided", () => {
            const event = {
                pathParameters: {
                    serviceId: "abc",
                },
            } as unknown as APIGatewayEvent;

            expect(() => getQueryInput(event)).toThrowError("Provided service ID is not valid");
        });
        it("throws a ClientError if invalid busStopType provided", () => {
            const event = {
                pathParameters: {
                    serviceId: "234",
                    busStopType: "invalid",
                },
            } as unknown as APIGatewayEvent;

            expect(() => getQueryInput(event)).toThrowError("Invalid bus stop type provided");
        });
    });

    describe("format service", () => {
        it("correctly formats db response", async () => {
            const formattedService: { outbound: ServiceStop[]; inbound: ServiceStop[] } = await formatStopsRoutes(
                stopsDbData,
            );

            expect(formattedService).toMatchSnapshot();
        });
    });
});
