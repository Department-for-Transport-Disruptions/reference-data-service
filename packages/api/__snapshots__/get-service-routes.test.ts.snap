// Vitest Snapshot v1

exports[`get-service-routes > format service > correctly formats db response 1`] = `
{
  "inbound": [],
  "outbound": [],
}
`;
